src/R2DH.jl

```diff
@@ -0,0 +1,310 @@
+    set_solver_specific!(stats, :Time_hist, outdict[:Time_hist])
+    set_solver_specific!(stats, :NonSmooth, outdict[:NonSmooth])
+    set_solver_specific!(stats, :SubsolverCounter, outdict[:Chist])
+    return stats
+  end
+
+function R2DH(
+  f::F,
+  ∇f!::G,
+  h::H,
+  D::DQN,
+  options::ROSolverOptions{R},
+  x0::AbstractVector{R};
+  Mmonotone::Int = 5,
+  selected::AbstractVector{<:Integer} = 1:length(x0),
+  kwargs...,
+) where {F <: Function, G <: Function, H, R <: Real, DQN <: AbstractDiagonalQuasiNewtonOperator}
+  start_time = time()
+  elapsed_time = 0.0
+  ϵ = options.ϵa
+  ϵr = options.ϵr
+  neg_tol = options.neg_tol
+  verbose = options.verbose
+  maxIter = options.maxIter
+  maxTime = options.maxTime
+  σmin = options.σmin
+  σk = options.σk
+  η1 = options.η1
+  η2 = options.η2
+  ν = options.ν
+  γ = options.γ
+  θ = options.θ
+
+  local l_bound, u_bound
+  has_bnds = false
+  for (key, val) in kwargs
+    if key == :l_bound
+      l_bound = val
+      has_bnds = has_bnds || any(l_bound .!= R(-Inf))
+    elseif key == :u_bound
+      u_bound = val
+      has_bnds = has_bnds || any(u_bound .!= R(Inf))
+    end
+  end
+
+  if verbose == 0
+    ptf = Inf
+  elseif verbose == 1
+    ptf = round(maxIter / 10)
+  elseif verbose == 2
+    ptf = round(maxIter / 100)
+  else
+    ptf = 1
+  end
+
+  # initialize parameters
+  xk = copy(x0)
+  hk = h(xk[selected])
+  if hk == Inf
+    verbose > 0 && @info "R2DH: finding initial guess where nonsmooth term is finite"
+    prox!(xk, h, x0, one(eltype(x0)))
+    hk = h(xk[selected])
+    hk < Inf || error("prox computation must be erroneous")
+    verbose > 0 && @debug "R2DH: found point where h has value" hk
+  end
+  hk == -Inf && error("nonsmooth term is not proper")
+
+  xkn = similar(xk)
+  s = zero(xk)
+  ψ = has_bnds ? shifted(h, xk, l_bound - xk, u_bound - xk, selected) : shifted(h, xk)
+
+  Fobj_hist = zeros(maxIter+1)
+  Hobj_hist = zeros(maxIter+1)
+  time_hist = zeros(maxIter+1)
+  FHobj_hist = fill!(Vector{R}(undef, Mmonotone), R(-Inf))
+  Complex_hist = zeros(Int, maxIter+1)
+  k_prox = 0
+  if verbose > 0
+    #! format: off
+    @info @sprintf "%6s %8s %8s %7s %8s %7s %7s %7s %1s" "iter" "f(x)" "h(x)" "√(ξ/ν)" "ρ" "σ" "‖x‖" "‖s‖" ""
+    #! format: off
+  end
+
+  local ξ
+  k = 0
+
+  fk = f(xk)
+  ∇fk = similar(xk)
+  ∇f!(∇fk, xk)
+  ∇fk⁻ = copy(∇fk) 
+  spectral_test = isa(D, SpectralGradient)
+ # D.d .= D.d .+ σk
+  Dkσk = D.d .+ σk
+  DNorm = norm(D.d, Inf)
+
+  ν = 1 / ((DNorm + σk) * (1 + θ))
+  mν∇fk = -ν * ∇fk
+  sqrt_ξ_νInv = one(R)  
+
+  optimal = false
+  tired = maxIter > 0 && k ≥ maxIter || elapsed_time > maxTime
+
+  while !(optimal || tired)
+    k_prox += 1
+    # model with diagonal hessian
+    φ(d) = ∇fk' * d + (d' * (Dkσk .* d)) / 2
+    mk(d) = φ(d) + ψ(d)
+
+    if spectral_test
+      prox!(s, ψ, mν∇fk, ν)
+    else
+      iprox!(s, ψ, ∇fk, Dkσk)
+    end
+    mks = mk(s)
+
+    if mks < -1e5
```

Comment:
What is this for???

Reply:
This is triggered when $ m_{ks} $ is very small (approaching $ -\infty $). In such cases, we do not compute $ f(x_{kn}) $, and we simply assume that the iteration is unsuccessful.

Reply:
Do you think that $1e-5$ is not that small ?

Reply:
I don’t understand why we would need that. Firstly, -1e5 is completely arbitrary. And secondly, why would the iteration be unsuccessful if the model decrease is large?

Reply:
Because of this 
<img width="633" alt="Condition-rhok=0" src="https://github.com/user-attachments/assets/05c950c2-080b-4819-b6e5-ec396cce2c17">

$$
\rho_k = 0
$$
Hence the iteration is automatically uncessuful

Reply:
However, I agree that -1e5 might be somewhat arbitrary. Should we set it to -Inf instead? 
We should also ensure that the ShiftedProximalOperators handle this appropriately when encountering \(d[i] < 0\). 

My concern is that Julia might struggle with calculus involving -Inf and interpret it as NaN. However, this does not seem to be the case, except in scenarios like \(\text{Inf} - \text{Inf}\) or \(\frac{\text{Inf}}{\text{Inf}}\), which yield NaN. In the paper, the latter, which may be \(\rho\), should be equal to 0 by convention, but Julia treats it as NaN.

Reply:
It should be -Inf, and ShiftedProximalOperators should return -Inf.